snuba/consumers/consumer.py

```diff
@@ -355,6 +360,77 @@ def build_writer() -> ProcessedMessageBatchWriter:
     return build_writer
 
 
+class DeadLetterStep(
+    ProcessingStep[Tuple[StorageKey, Union[None, BytesInsertBatch, ReplacementBatch]]]
+):
+    """
+    If a batch failed to be inserted to ClickHouse for some reason, we put those messages
+    into a dead letter topic so that they can be consumed later and tried again.
+
+    This is only possible for the MultiStorageConsumerFactory right now, and it is an
+    optional step for the MultistorageCollector. Additionally, this is only enabled if
+    the storages have `ignore_errors` set to True - which is the case for the errors_v2
+    and transaction_v2 storages.
+
+    """
+
+    def __init__(self, producer: AbstractProducer[KafkaPayload], topic: Topic):
+        self.__producer = producer
+        self.__topic = topic
+        self.__futures: Deque[Future[Message[KafkaPayload]]] = deque()
+        self.__closed = False
+
+    def __format_payload(
+        self,
+        message: Message[
+            Tuple[StorageKey, Union[None, BytesInsertBatch, ReplacementBatch]]
+        ],
+    ) -> List[KafkaPayload]:
+        kafka_payloads: List[KafkaPayload] = []
+        storage_key, payload = message.payload
+        if isinstance(payload, BytesInsertBatch):
+            for row in payload.rows:
+                kafka_payloads.append(
+                    KafkaPayload(storage_key.value.encode("utf-8"), row, [])
+                )
+        return kafka_payloads
+
+    def poll(self) -> None:
+        pass
+
+    def submit(
+        self,
+        message: Message[
+            Tuple[StorageKey, Union[None, BytesInsertBatch, ReplacementBatch]]
+        ],
+    ) -> None:
+        payloads = self.__format_payload(message)
+        for payload in payloads:
+            future = self.__producer.produce(self.__topic, payload)
+            self.__futures.append(future)
+
+    def close(self) -> None:
+        self.__closed = True
+
+    def terminate(self) -> None:
+        self.__closed = True
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        start = time.time()
+
+        while self.__futures:
+            if not timeout or timeout > time.time() - start:
+                break
```

Comment:
An edge case which can happen here is that the producer is not done writing the failed insert message to the dead letter topic before the timeout is complete. In those cases we will lose data since the consumer will commit the kafka offset. The question then is, whether we can live with that or should we prevent committing the kafka offset?

Reply:
Hmm so @nikhars are you suggesting we raise here if we timeout? there is also the case where soemthing failed when producing (I just log a warning below for now). We could log a warning for timeouts as well for now first? Not sure at what point we want to stop ingestion

Reply:
Can we add metrics around how many times we are failing to produce. Either because of timeouts or problems in futures. 

I agree we can not do anything more than that for now.